android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentService.java

```diff
@@ -0,0 +1,876 @@
+            try {
+                messenger.send(Message.obtain(null, LIFECYCLE_MSG_EXIT_CODE, exitCode, 0));
+            } catch (RemoteException | IllegalStateException e) {
+                // probably an observer has crashed or terminated, remove from the set
+                it.remove();
+                Log.w(LOG_TAG, "Unable to notify remote node " + Log.getStackTraceString(e));
+            }
+        }
+
+        stopForeground(true);
+        stopSelf();
+        setStopped();
+    }
+
+    private void terminateThread(Thread thread, boolean join) {
+        Log.d(LOG_TAG, "terminateThread");
+        if (thread != null && thread.isAlive()) {
+            thread.interrupt();
+            if (join) {
+                try {
+                    thread.join(threadJoinTime.get());
+                } catch (InterruptedException e) {
+                    Log.e(LOG_TAG, "Exception " + Log.getStackTraceString(e));
+                }
+            }
+        }
+    }
+
+    private String getMasterPackage() {
+        return TextUtils.isEmpty(masterPackage) ? NUCLEUS_PACKAGE : masterPackage;
+    }
+
+    private void registerBroadcastReceiver() {
+        registerReceiver(receiver, new IntentFilter(ACTION_STOP_COMPONENT));
+    }
+
+    private class MessengerLooperThread extends Thread {
+        private IncomingSignalsHandler inSigHandler = null;
+        private AtomicReference<Messenger> inSigMessenger = new AtomicReference<>();
+        private Looper inSigLooper = null;
+
+        @Override
+        public void run() {
+            Looper.prepare();
+            synchronized (inSigMessenger) {
+                inSigLooper = Looper.myLooper();
+                inSigHandler = new IncomingSignalsHandler(inSigLooper);
+                inSigMessenger.set(new Messenger(inSigHandler));
+                inSigMessenger.notifyAll();
+            }
+
+            Looper.loop();
+        }
+
+        public IBinder getMessengerBinder() {
+            synchronized (inSigMessenger) {
+                if (inSigMessenger.get() == null) {
+                    try {
+                        inSigMessenger.wait(MESSENGER_READINESS_WAIT_LIMIT_MS);
+                    } catch (InterruptedException ignore) {
+                        Log.d(LOG_TAG, "Interrupted while waiting for the messenger to be ready");
+                    }
+                }
+                Messenger messenger = inSigMessenger.get();
+                if (messenger != null) {
+                    return messenger.getBinder();
+                } else {
+                    return null;
+                }
+            }
+        }
+
+        void finish() {
+            if (this.isAlive()) {
+                // Terminate looper
+                inSigLooper.quitSafely();
+                // Wait for the looper top process all the messages in its queue
+                try {
+                    this.join();
+                } catch (InterruptedException ignore) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)